src/hotspot/share/opto/library_call.cpp

```diff
@@ -6967,6 +6969,77 @@ bool LibraryCallKit::inline_base64_decodeBlock() {
+  assert(ainst != NULL, "Accumulator obj is null");
+  assert(ainst->is_loaded(), "MutableIntegerModuloP obj is not loaded");
+  ciKlass* klass_MutableElement = ainst->instance_klass()->find_klass(ciSymbol::make("sun/security/util/math/intpoly/IntegerPolynomial$MutableElement"));
+  assert(klass_MutableElement != NULL, "IntegerPolynomial$MutableElement class is null");
+  assert(klass_MutableElement->is_loaded(), "IntegerPolynomial$MutableElement class is not loaded");
+  ciInstanceKlass* instklass_MutableElement = klass_MutableElement->as_instance_klass();
+
+  const TypeKlassPtr* aklass = TypeKlassPtr::make(instklass_MutableElement);
+  const TypeOopPtr* atype = aklass->as_instance_type()->cast_to_ptr_type(TypePtr::NotNull);
+  Node* accObj = new CheckCastPPNode(control(), accFace, atype);
+  accObj = _gvn.transform(accObj);
+  Node* alimbs = load_field_from_object(accObj, "limbs", "[J");
+
+  Node* rFace = load_field_from_object(polyObj, "r", "Lsun/security/util/math/IntegerModuloP;"); //this.r.limbs
+  assert(rFace != NULL, "R field is null");
+  const TypeInstPtr* rinst = _gvn.type(rFace)->isa_instptr();
+  assert(rinst != NULL, "R obj is null");
+  assert(rinst->is_loaded(), "IntegerModuloP obj is not loaded");
+  ciKlass* klass_ImmutableElement = rinst->instance_klass()->find_klass(ciSymbol::make("sun/security/util/math/intpoly/IntegerPolynomial$ImmutableElement"));
+  assert(klass_ImmutableElement != NULL, "IntegerPolynomial$ImmutableElement class is null");
+  assert(klass_ImmutableElement->is_loaded(), "IntegerPolynomial$ImmutableElement class is not loaded");
+  ciInstanceKlass* instklass_ImmutableElement = klass_ImmutableElement->as_instance_klass();
+
+  const TypeKlassPtr* rklass = TypeKlassPtr::make(instklass_ImmutableElement);
+  const TypeOopPtr* rtype = rklass->as_instance_type()->cast_to_ptr_type(TypePtr::NotNull);
+  Node* rObj = new CheckCastPPNode(control(), rFace, rtype);
+  rObj = _gvn.transform(rObj);
+  Node* rlimbs = load_field_from_object(rObj, "limbs", "[J");
+
+  input = must_be_not_null(input, true);
+  alimbs = must_be_not_null(alimbs, true);
+  rlimbs = must_be_not_null(rlimbs, true);
+
+  // Intrinsic assumes there are exactly 5 limbs! Currently enforced by IntegerModuloP.checkLimbsForIntrinsic
+  // FIXME: where to branch to if limbs array length != 5? Could be an 'assert'/RuntimeException
+  // FIXME: repeat for rlimbs
+  // Node* cmp = _gvn.transform(new CmpINode(load_array_length(alimbs), intcon(5)));
+  // Node* bol = _gvn.transform(new BoolNode(cmp, BoolTest::eq));
+  // Node* if_eq = generate_slow_guard(bol, slow_region);
+
+  Node* input_start = array_element_address(input, input_offset, T_BYTE);
+  assert(input_start, "input array is NULL");
+  Node* acc_start = array_element_address(alimbs, intcon(0), T_LONG);
+  assert(acc_start, "acc array is NULL");
+  Node* r_start = array_element_address(rlimbs, intcon(0), T_LONG);
+  assert(r_start, "r array is NULL");
+
+  Node* call = make_runtime_call(RC_LEAF,
```

Comment:
Can we safely change this to `RC_LEAF | RC_NO_FP`?  For the ChaCha20 block intrinsic I'm working on I've been using that parameter because I'm not touching the FP registers and that looks to be the case here (though your intrinsic is a lot more complicated than mine so I may have missed something).  I believe the GHASH and AES library call routines also call `make_runtime_call()` in this way.

Reply:
Makes sense to me, will put it in and re-test (no fp registers anywhere in the intrinsic). Thanks!

Reply:
done